EngineLayer/Proteomics/Protease.cs

```diff
@@ -87,9 +93,337 @@ internal List<int> GetDigestionSiteIndices(string sequence)
+                        intervals.Add(new ProteinInterval(proteolysisProduct.OneBasedBeginPosition.Value, proteolysisProduct.OneBasedEndPosition.Value, protein, 0, proteolysisProduct.Type));
+                    }
+                }
+            }
+
+            // Cleavage rules for nonspecific search
+            else if (CleavageSpecificity == CleavageSpecificity.Semi)
+            {
+                intervals.AddRange(SemiProteolyticDigestion(protein, initiatorMethionineBehavior, maximumMissedCleavages, minPeptidesLength, maxPeptidesLength));
+            }
+
+            // Full proteolytic cleavage
+            else if (CleavageSpecificity == CleavageSpecificity.Full)
+            {
+                intervals.AddRange(FullDigestion(protein, initiatorMethionineBehavior, maximumMissedCleavages, minPeptidesLength, maxPeptidesLength));
+            }
+
+            else
+            {
+                throw new NotImplementedException();
+            }
+
+            return intervals;
+        }
+
+        /// <summary>
+        /// Gets the protein intervals based on nonspecific digestion rules
+        /// </summary>
+        /// <param name="protein"></param>
+        /// <param name="initiatorMethionineBehavior"></param>
+        /// <param name="maximumMissedCleavages"></param>
+        /// <param name="minPeptidesLength"></param>
+        /// <param name="maxPeptidesLength"></param>
+        /// <returns></returns>
+        private IEnumerable<ProteinInterval> SemiProteolyticDigestion(Protein protein, InitiatorMethionineBehavior initiatorMethionineBehavior, int maximumMissedCleavages, int? minPeptidesLength, int? maxPeptidesLength)
+        {
+            List<ProteinInterval> intervals = new List<ProteinInterval>();
+            List<int> oneBasedIndicesToCleaveAfter = GetDigestionSiteIndices(protein.BaseSequence);
+            for (int i = 0; i < oneBasedIndicesToCleaveAfter.Count - maximumMissedCleavages - 1; i++)
+            {
+                bool retain = Retain(i, initiatorMethionineBehavior, protein[0]);
+                bool cleave = Cleave(i, initiatorMethionineBehavior, protein[0]);
+                if (!retain && !cleave)
+                {
+                    continue;
+                }
+
+                int nTerminusProtein = retain ? oneBasedIndicesToCleaveAfter[i] : 1;
+                int cTerminusProtein = oneBasedIndicesToCleaveAfter[i + maximumMissedCleavages + 1];
+                if (OkayLength(cTerminusProtein - nTerminusProtein, minPeptidesLength, maxPeptidesLength))
+                {
+                    intervals.Add(new ProteinInterval(nTerminusProtein + 1, cTerminusProtein, protein, cTerminusProtein - nTerminusProtein, "semi" + (cleave ? ":M cleaved" : ""))); // Maximum sequence length
+                }
+
+                // Fixed termini at each internal index
+                IEnumerable<int> internalIndices = Enumerable.Range(nTerminusProtein - 1, cTerminusProtein - nTerminusProtein - 1);
```

Comment:
@zrolfs You were looping through the same indices in forward and reverse directions with the `for` loops... more efficient just to do it once, here with Linq.